docs/customization/plugin-authors-guide.md

```diff
@@ -250,12 +250,209 @@ kaleidoscope::plugin::MyPlugin;
 
 In the above example, the private member variable `start_time_` and the constant `timeout` are the same type of unsigned integer (`uint16_t`), and we've used the additional boolean `timer_running_` to keep from checking for timeouts when `start_time_` isn't valid.  This plugin does something (unspecified) 500 milliseconds after a `Key_X` toggles on.
 
-## Creating artificial events
+## Creating additional events
 
-## Controlling LEDs
+Another thing we might want a plugin to do is generate "extra" events that don't correspond to physical state changes.  An example of this is the Macros plugin, which might turn a single keypress into a series of HID reports sent to the host.  Let's build a simple plugin to illustrate how this is done, by making a key type a string of characters, rather than a single one.
 
-## HID reports
+For the sake of simplicity, let's make the key `H` result in the string `Hi!` being typed (from the point of view of the host computer).  To do this, we'll make a plugin with an `onKeyEvent()` handler (because we want it to respond to a particular keypress event), which will call `Runtime.handleKeyEvent()` to generate new events sent to the host.
+
+The first thing we need to understand to do this is how to use the `KeyEvent()` constructor to create a new `KeyEvent` object.  For example:
+
+```c++
+KeyEvent event = KeyEvent(KeyAddr::none(), IS_PRESSED, Key_H);
+```
+
+This creates a `KeyEvent` where `event.addr` is an invalid address that doesn't correspond to a physical keyswitch, `event.state` has only the `IS_PRESSED` bit set, but not `WAS_PRESSED`, which corresponds to a key toggle-on event, and `event.key` is set to `Key_H`.
+
+We can then cause Kaleidoscope to process this event, including calling plugin handlers, by calling `Runtime.handleKeyEvent(event)`:
+
+```c++
+  EventHandlerResult onKeyEvent(KeyEvent &event) {
+    if (event.key == Key_H && keyToggledOn(event.state)) {
+
+	  // Create and send the `H` (shift + h)
+	  KeyEvent new_event = KeyEvent(KeyAddr::none(), IS_PRESSED, LSHIFT(Key_H));
+	  Runtime.handleKeyEvent(new_event);
+
+	  // Change the key value and send `i`
+      new_event.key = Key_I;
```

Comment:
There are some indentation issues, also in the code snippets below.

Reply:
Thanks.  I'll go through it and fix them up.

Reply:
Ah, I see the problem.  I was writing all that in markdown-mode, and tab characters ended up where I didn't expect them.